app/lib/features/room/actions/join_room.dart

```diff
@@ -14,54 +16,68 @@ Future<T> _ensureLoadedWithinTime<T>(
   Future<T?> Function() callback, {
   int delayMs = 300,
   int attempts = 20,
+  bool throwError = false,
 }) async {
   int remaining = attempts;
   while (remaining > 0) {
     remaining -= 1;
-    final res = await callback();
-    if (res != null) {
-      return res;
+    try {
+      final res = await callback();
+      if (res != null) {
+        return res;
+      }
+    } catch (e) {
+      if (throwError) rethrow;
     }
     await Future.delayed(Duration(milliseconds: delayMs));
   }
 
   throw 'Loading timed out';
 }
 
-Future<String?> joinRoom(
-  BuildContext context,
-  WidgetRef ref,
-  String displayMsg,
-  String roomIdOrAlias,
-  String? server,
-  Function(String)? forward,
-) async {
+Future<String?> joinRoom({
+  required BuildContext context,
+  required WidgetRef ref,
+  required String roomIdOrAlias,
+  required List<String>? serverNames,
+  String? displayMsg,
+  String? roomName,
+
+  /// configure to throw on error rather than return null
+  bool throwOnError = false,
+}) async {
   final lang = L10n.of(context);
-  EasyLoading.show(status: displayMsg);
-  final client = ref.read(alwaysClientProvider);
+  EasyLoading.show(
+    status: displayMsg ?? lang.tryingToJoin(roomName ?? roomIdOrAlias),
+  );
   try {
-    final newRoom = await client.joinRoom(roomIdOrAlias, server);
+    final client = ref.read(alwaysClientProvider);
+    final sdk = await ref.read(sdkProvider.future);
+    VecStringBuilder servers = sdk.api.newVecStringBuilder();
+    if (serverNames != null) {
+      for (final server in serverNames) {
+        servers.add(server);
+      }
+    }
+    final newRoom = await client.joinRoom(roomIdOrAlias, servers);
     final roomId = newRoom.roomIdStr();
-    final isSpace = newRoom.isSpace();
-    // ensure we re-evaluate the room data on our end. This is necessary
-    // if we knew of the room prior (e.g. we had left it), but hadn’t joined
-    // this should properly re-evaluate all possible readers
-    ref.invalidate(maybeRoomProvider(roomId));
+    final isSpace = await _ensureLoadedWithinTime(() async {
+      final room = await ref.refresh(maybeRoomProvider(roomId).future);
+      return room?.isJoined() == true ? room!.isSpace() : null;
+    });
```

Comment:
this now properly checks for the room being available _and_ accessible to the UI before continuing.